tutorials/super_blink/research/adc.c

```diff
@@ -0,0 +1,87 @@
+#include "libs/adc/api.h"
+
+#include <avr/interrupt.h>
+#include <avr/io.h>
+#include <stdlib.h>
+
+void (*interrupt_callback)(void) = NULL;
+
+ISR(ADC_vect) {
+    if (interrupt_callback != NULL) {
+        (*interrupt_callback)();
+    }
+}
+
+/*
+ * Initialize the adc on the MCU
+ *
+ * ADCSRA = ADC register A containing configuration bits (ADC Control and Status
+ * Register A)
+ *
+ * ADCSRB = ADC register B containing more configuration bits
+ *
+ * ADEN = ADC enable bit
+ *
+ * ADPS2:0 = ADC prescaler select - division factor between
+ * system clock fq and input clock to ADC
+ *
+ * AREFEN = Analog reference pin enable
+ *
+ * ADMUX = Register for muxing to various ADC inputs
+ *
+ * REFS0 = REFSn: Selects reference for the ADC
+ */
+
+// Initializes the ADC on the MCU
+void adc_init(void) {
+    ADCSRA = (1 << ADEN) | (1 << ADPS2)
+             | (1 << ADPS0); // Sets the ADEN, ADPS2, and ADPS0 bits to 1.
+                             // Enables the ADC, sets the prescaler to 32
+    ADCSRB &= (1 << AREFEN); // What does this do? Clear all bits but AREFEN?
+    ADMUX |= (1 << REFS0); // Bits: 01000000; ADC reference voltage is AV_CC
+}
+
+// Starts ADC converting on a pin
+void adc_start_convert(
+    adc_pin_e pin) { // Takes a type adc_pin_e named pin, maps back to an enum
+    ADMUX &= ~0x1F; // In binary: 11100000. Clears all bits but the 3 MSBs
+    // Clears bits MUX[4:0] in prep for setting the device to convert
+    ADMUX |= (pin & 0x1F); // Sets the pin/device to convert
+    ADCSRA |= (1 << ADSC); // Starts conversion
+}
+
+// Check if an ADC conversion is done, and if so reads the result and sets it to
+// result
+int adc_poll_complete(uint16_t* result) { // Takes a pointer to a 16 bit int
+    // Checks if the ADIF bit is 0
+    if (!(ADCSRA & (1 << ADIF))) {
+        return -1;
+    } else { // If ADIF is 1
+        ADCSRA |= 1 << ADIF; // If ADIF is already 1, why do we need this?
```

Comment:
On line 60 - this sets the `ADIF` bit to 1, correct? Based on the conditional this seems redundant.

Reply:
Not quite sure on this one either to be honest - looking at the datasheet, it says that the ADIF is automatically cleared by the hardware, or can be alternately cleared by manually setting it to 1 (so somewhat counterintuitively, you write a 1 to set it to 0, also called write to clear). The reason why they do this is a little complex, but long story short it protects you from something called read-modify-write errors - essentially having the value of the register changing from under you while you're trying to clear a bit means you can accidentally clear stuff you didn't mean to. (don't worry if that didn't make total sense; I'm probably doing a bad job explaining it: [this](https://www.avrfreaks.net/s/topic/a5C3l000000UDgREAW/t066397) link has some more info)

Now, why we're manually clearing it I'm not too sure. It's possible the auto clearing wasn't working totally right (maybe because we don't have an interrupt attached?), as if the flag isn't cleared it looks like it won't let you run another conversion, so when we wrote the library the fix was to manually clear. Would be interesting to run a test without this line and see what happens!